### PR TITLE
change output format to json-patch

### PR DESCRIPTION
**Pull request by harunurhan**
TODO:

- [x]  Tests
- [x]  Update docs

### CODE REVIEW

**Comment by jmartinm** [comment]
Looks good to me. Just the items in the `TODO` to be done.

File: json_merger/conflict.py [hunk 1]
```diff
@@ -93,5 +94,25 @@ def with_prefix(self, root_path):
         return Conflict(self.conflict_type, root_path + self.path, self.body)
 
     def to_json(self):
-        """Deserializes conflict to a JSON object."""
-        return json.dumps([self.conflict_type, self.path, self.body])
+        """Deserializes conflict to a JSON object with json-patch format"""
+        # map ConflictType to json-patch operator
+        if (
```
Comment (jmartinm) [inline json_merger/conflict.py]: You can do this in one line with:

```python
if self.conflict_type in ('REORDER', 'SET_FIELD'):
```

Same for the next `elif`

File: json_merger/conflict.py [hunk 2]
```diff
@@ -93,5 +94,27 @@ def with_prefix(self, root_path):
...
+        """Deserializes conflict to a JSON object.
+        
+        It returns `json-patch <https://tools.ietf.org/html/rfc6902>`_ format.
+            - REORDER, SET_FIELD become "op": "replace"
+            - MANUAL_MERGE, ADD_BACK_TO_HEAD become "op": "add"
+            - Path array becomes `json-pointer 
+            <https://tools.ietf.org/html/rfc6901>`_
+            - Original conflict type is added to "$type"
+        """
+        # map ConflictType to json-patch operator
```
Comment (jmartinm) [inline json_merger/conflict.py]: This can go in the docstring instead of comment.

File: tests/unit/test_conflict.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    assert json.loads(conflict_json) == {
+        '$type': 'ADD_BACK_TO_HEAD',
+        'op': 'add',
+        'path': '/foo/bar',
+        'value': {}
+    }
+
+
+def test_to_json_with_remove_field():
+    conflict = Conflict('REMOVE_FIELD', ['foo', 'bar'], None)
```
Comment (jmartinm) [inline tests/unit/test_conflict.py:80]: Value sometimes is `None` and sometimes empty dictionary?

File: json_merger/conflict.py [hunk 4]
```diff
@@ -93,5 +94,27 @@ def with_prefix(self, root_path):
...
+        if self.conflict_type in ('REORDER', 'SET_FIELD'):
+            op = 'replace'
+        elif self.conflict_type in ('MANUAL_MERGE', 'ADD_BACK_TO_HEAD'):
+            op = 'add'
+        else:
+            op = 'remove'
+        # stringify path array
+        path = '/' + '/'.join(str(el) for el in self.path)
```
Comment (jmartinm) [inline json_merger/conflict.py]: Why is `str` needed? path has numbers for array items? You can add a test case that shows this case.

**Comment by harunurhan** [comment]
when is the operation is `add`, the `path` should point to an index to be inserted or `-` to append at the end. So we need to `-` or `0` to path in order to produce correct `json-patch`. Which one makes more sense for you @jmartinm @michamos

File: tests/unit/test_conflict.py (hunk 3 above)
Comment (harunurhan) [reply tests/unit/test_conflict.py:80]: Empty dictionary is to represent a value, but for `REMOVE_FIELD` conflict, there is no value.

File: json_merger/conflict.py (hunk 4 above)
Comment (harunurhan) [reply json_merger/conflict.py]: I am not sure if path has a `number` or a `string` for `index`, I am gonna check the examples or try to reproduce it but maybe @kaplun knows?

So if it doesn't we can remove this conversion.

Comment (kaplun) [reply json_merger/conflict.py]: I miss a bit of the context, but if I am understanding correctly indeed array positions are very likely integers.

File: json_merger/conflict.py [hunk 5]
```diff
@@ -93,5 +94,27 @@ def with_prefix(self, root_path):
...
+        - Path becomes `json-pointer <https://tools.ietf.org/html/rfc6901>`_
+        - Original conflict type is added to "$type"
+        """
+        # map ConflictType to json-patch operator
+        if self.conflict_type in ('REORDER', 'SET_FIELD'):
+            op = 'replace'
+        elif self.conflict_type in ('MANUAL_MERGE', 'ADD_BACK_TO_HEAD'):
+            op = 'add'
+        else:
+            op = 'remove'
```
Comment (kaplun) [inline json_merger/conflict.py]: Isn't this a bit a strong default behavior? Can you be more explicit in which operations and up in `remove` and raise an exception for unknown operations, just to be future-safe?

File: json_merger/conflict.py (hunk 4 above)
Comment (harunurhan) [reply json_merger/conflict.py]: Yes, actually I found an example and they are integers so I am gonna add a test case where path has integer as @jmartinm suggested

File: json_merger/conflict.py (hunk 5 above)
Comment (harunurhan) [reply json_merger/conflict.py]: @kaplun 
I was thinking the same but then in the `__new__` of `Conflict` we already raise an error if it's not one of these.

However if you think it's better I can raise similar here in `else` and make `remove` case explicit.

Comment (kaplun) [reply json_merger/conflict.py]: Yeah thanks! That would make the code more explicit and safe indeed.

**Comment by harunurhan** [comment]
Addressed all comments, only thing to do is https://github.com/inveniosoftware-contrib/json-merger/pull/29#issuecomment-284703338

**Comment by kaplun** [comment]
@harunurhan that depends on the policy of the rule. It makes quite a difference in the metadata to append at the end or insert at the beginning. Isn't there some more info available (e.g. on the documentation) about what was the intended semantic of `add`?

**Comment by michamos** [comment]
I am not quite sure I understand the meaning of the different conflicts well-enough to be convinced that the translation to JSON patches is correct. As this is not really high priority, maybe it can wait till we start working on the merger configuration?

**Comment by kaplun** [comment]
So you propose @harunurhan simply assumes there's going to be patches however produced and keep on working from that?

**Review by ammirate — APPROVED** [review]
Looks good to me

**Pull request merged**